core/include/core/detector.hpp

```diff
@@ -105,38 +107,111 @@ namespace detray
          * 
          * @tparam portal_type the type of the portal mask
          * 
+         * @param volume_index the volume index to which this portal should be added
          * @param transform the transform of the portal surface
          * @param portals the vector of portal masks
-         * @param volume to which this portal belongs to
+         * 
+         * @returns the index of the newly added portal surface
          **/
         template <typename portal_type>
-        void add_portal_surface(transform_type &&transform, const dvector<portal_type> &portals, volume &volume)
+        dindex 
+        add_portal_surface(dindex volume_index, transform_type &&transform, const dvector<portal_type> &portals)
         {
+            auto &volume = _volumes[volume_index];
             // Get the boundary group, record the index and insert the portals
             auto &group = std::get<dvector<portal_type>>(_portal_masks);
-            guaranteed_index index_start = group.size();
-            guaranteed_index index_end = static_cast<guaranteed_index>(index_start + portals.size() - 1);
+            dindex index_start = group.size();
+            dindex index_end = static_cast<dindex>(index_start + portals.size() - 1);
             group.insert(group.end(), portals.begin(), portals.end());
             // Create a range of portal masks
-            guaranteed_range range = {index_start, index_end};
-            guaranteed_index type = _portal_types.find(portal_type::mask_identifier)->second;
+            dindex_range range = {index_start, index_end};
+            dindex type = _portal_types.find(portal_type::mask_identifier)->second;
             // Record the portal index
             volume.portal_surface_indices.push_back(_portal_surfaces.size());
-            guaranteed_index volume_index = volume.index;
-            typed_guaranteed_range links = {type, range};
+            typed_dindex_range links = {type, range};
             // Record the transform index
-            guaranteed_index transform_index = _portal_transforms.size();
+            dindex transform_index = _portal_transforms.size();
             _portal_transforms.push_back(std::move(transform));
             _portal_surfaces.push_back(portal_surface(std::move(transform_index), std::move(links), std::move(volume_index), false));
+            return transform_index;
+        }
+
+        /** Update portal links
+         *
+         * @param portal_mask The portal mask which will be updated
+         * @param additional_link The new link to be added to the portal
+         *
+         */
+        template <typename mask_type>
+        void update_portal_links(mask_type& mask, portal_links additional_link) noexcept(false) {
+            auto &mask_link = mask.links();
+            for (dindex il = 0; il < mask_link.size(); ++il){
+                auto& link  = mask_link[il];
+                auto& new_link = additional_link[il];
```

Comment:
Is it guaranteed that these two collections always have the sam size?

Reply:
It should - otherwise the reuse doesn't make sense ... what I have in mind is to do as little checking as possible at runtime, but you're right we should just throw exceptions when building.

Reply:
Fixed that.